File: hr-v4-api/src/main/java/com/globits/hr/service/CandidateRecruitmentRoundService.java
package com.globits.hr.service;

import com.globits.core.service.GenericService;
import com.globits.hr.HrConstants;
import com.globits.hr.domain.Candidate;
import com.globits.hr.domain.CandidateRecruitmentRound;
import com.globits.hr.dto.ApiResponse;
import com.globits.hr.dto.CandidateDto;
import com.globits.hr.dto.CandidateRecruitmentRoundDto;
import com.globits.hr.dto.RecruitmentDto;
import com.globits.hr.dto.search.SearchCandidateDto;
import com.globits.hr.dto.search.SearchCandidateRecruitmentRoundDto;
import com.globits.hr.dto.search.SearchRecruitmentDto;
import com.globits.salary.dto.SalaryResultDto;
import com.globits.salary.dto.search.SearchSalaryResultDto;
import org.springframework.data.domain.Page;

import java.util.HashMap;
import java.util.List;
import java.util.UUID;

public interface CandidateRecruitmentRoundService extends GenericService<CandidateRecruitmentRound, UUID> {
    void handleSetCandidateRecruitmentRoundList(CandidateDto dto, Candidate entity);

    Boolean isValid(CandidateRecruitmentRoundDto dto);

    CandidateRecruitmentRoundDto saveCandidateRecruitmentRound(CandidateRecruitmentRoundDto dto);
    CandidateRecruitmentRound convertDtoToEntity(CandidateRecruitmentRoundDto dto);

    Page<CandidateRecruitmentRoundDto> pagingCandidateRecruitmentRound(SearchCandidateRecruitmentRoundDto dto);

    CandidateRecruitmentRoundDto getById(UUID id);

    Boolean remove(UUID id);

    Boolean removeMultiple(List<UUID> ids);

    // update candidate's result in recruitment round
    Boolean updateCandidateRecruitmentRoundResult(SearchCandidateRecruitmentRoundDto dto) throws Exception;

    // Phân bổ/Sắp xếp ứng viên cho các vòng tuyển dụng tiếp theo
    Boolean moveToNextRecruitmentRound(SearchCandidateRecruitmentRoundDto dto) throws Exception;

    // Phân bổ/Sắp xếp ứng viên cho vòng tuyển dụng đầu tiên
    Boolean distributeCandidatesForFirstRecruitmentRound(SearchCandidateRecruitmentRoundDto dto) throws Exception;

    Boolean checkExistCandidateRecruitmentRound(UUID candidateId, UUID recruitmentRoundId);

    ApiResponse<List<CandidateRecruitmentRoundDto>> getListCandiDateByPlainAndRound(UUID planId, UUID roundId);

    ApiResponse<Boolean> confirmInterview(UUID candidateId, HrConstants.CandidateRecruitmentRoundStatus desition);

    ApiResponse<Boolean> doActionAssignment(UUID crrId, HrConstants.CandidateStatus status);
//    ApiResponse<List<CandidateRecruitmentRoundDto>> approveCandidateRound(SearchCandidateRecruitmentRoundDto dto);

    ApiResponse<List<CandidateRecruitmentRoundDto>> getByIdRecruitmentRound(UUID roundId);

    ApiResponse<Boolean> passToNextRound(UUID crrId);

    ApiResponse<Boolean> rejectCandidateRound(UUID crrId);

    List<CandidateRecruitmentRoundDto> getCandidateRoundByCandidateId(UUID candidateId);

    ApiResponse<HashMap<UUID, HashMap<String, Object>>> passListToNextRound(List<UUID> crrIds);

    ApiResponse<Boolean> rejectListCandidateRound(List<UUID> crrIds);
}

File: hr-v4-client-app/src/app/views/Candidate/CandidateRecruitmentRound/CandidateRecruitmentRoundService.js
import axios from "axios";
import ConstantList from "app/appConfig";

const API_PATH = ConstantList.API_ENPOINT + "/api/candidate-recruitment-round";


export const pagingCandidateRecruitmentRound = (searchObject) => {
    var url = API_PATH + "/paging-candidate-recruitment-round";
    return axios.post(url, searchObject);
};

export const getById = (id) => {
    let url = API_PATH + "/" + id;
    return axios.get(url);
};

export const saveCandidateRecruitmentRound = (obj) => {
    let url = API_PATH + "/save-candidate-recruitment-round";
    return axios.post(url, obj);
};

export const deleteCandidateRecruitmentRound = (id) => {
    let url = API_PATH + "/" + id;
    return axios.delete(url);
};

export const deleteMultiple = (ids) => {
    const url = API_PATH + "/delete-multiple";
    return axios.post(url, ids);
};

// update candidate's result in recruitment round
export const updateRecruitmentRoundResult = (searchObject) => {
    const url = API_PATH + "/update-candidate-recruitment-round-result";
    return axios.post(url, searchObject);
}

// Phân bổ/Sắp xếp ứng viên cho các vòng tuyển dụng tiếp theo
export const moveToNextRecruitmentRound = (searchObject) => {
    const url = API_PATH + "/move-to-next-recruitment-round";
    return axios.post(url, searchObject);
}

// Phân bổ/Sắp xếp ứng viên cho vòng tuyển dụng đầu tiên
export const distributeCandidatesForFirstRecruitmentRound = (searchObject) => {
    const url = API_PATH + "/distribute-candidates-for-first-recruitment-round";
    return axios.post(url, searchObject);
}
export const doActionAssignment = (crrId, status) => {
    const url = API_PATH + `/do-action-assignment/${crrId}/${status}`;
    return axios.get(url);
}
export const passToNextRound = (crrId) => {
    const url = API_PATH + `/pass-to-next-round/${crrId}`;
    return axios.get(url);
}

export const passListToNextRound = (searchObject) => {
    const url = API_PATH + `/pass-list-to-next-round`;
    return axios.post(url, searchObject);
}

export const rejectCandidateRound = (crrId) => {
    const url = API_PATH + `/reject-candidate-round/${crrId}`;
    return axios.get(url);
}

export const getCandidateRoundByCandidateId = (candidateId) => {
    const url = API_PATH + "/get-round-by-candidate/" + candidateId;
    return axios.get(url);
}